sycl/doc/extensions/DataFlowPipes/data_flow_pipes.asciidoc

```diff
@@ -0,0 +1,650 @@
+
+[source,c++,Different pipes,linenums]
+----
+// Write to a pipe
+myQueue.submit([&](handler& cgh) {
+  auto read_acc = readBuf.get_access<access::mode::read>(cgh);
+
+  cgh.parallel_for<class foo>(range<1> { 1024 }, [=](id<1> idx) {
+    pipe<class some_pipe, int>::write( read_add[idx] );
+  });
+});
+
+// Read from a different pipe (probably by accident)
+myQueue.submit([&](handler& cgh) {
+  auto write_acc = writeBuf.get_access<access::mode::write>(cgh);
+
+  cgh.parallel_for<class bar>(range<1> { 1024 }, [=](id<1> idx) {
+    write_acc[idx] = pipe<class some_pipe, int>:read();
+  });
+});
+
+// Write to yet a third pipe (probably by accident)
+{
+  pipe<class some_pipe, int>::write( read_add[idx] );
+}
+
+// Write to a fourth pipe (probably by accident)
+{
+  pipe<class some_pipe, int>::write( read_add[idx] );
+}
+  
+// Forward declaring `class some_pipe` before this code block would have
+// made all writes access the same pipe type.  It is highly encouraged to define
+// pipe types through `using` once, at a scope visible to all uses, unless
+// C++ scoping is intentionally being used to create unique pipe types.
+----
+
+An outstanding issue is whether the code example above (with writes to independent pipes) should lead to a warning within implementations.  Backends will typically already error if a pipe doesn't have both read and write endpoint(s), but it's easy to generate code examples where this condition is met and multiple pipe types still lead to unexpected behavior (to a user not accustomed to {cpp} scoping rules).
+
+== Pipe types and type aliasing
+
+Type aliases in {cpp} through the `using` mechanism do not change the type of a pipe.  For example, the two writes in the following code snip are to the same pipe, even though _name_ in the second write is an alias to that used in the first write.
+
+[source,c++,Read write members,linenums]
+----
+  class a_class;
+  using type_alias = a_class;
+  pipe<a_class, int>::write(0);
+  pipe<type_alias, int>::write(0);
+----
+
+== Host pipe map/unmap
+
+Pipes expose two additional static member functions that are available within host code, and which map to the OpenCL C host pipe extension map/unmap interface.  These methods provide higher bandwidth or otherwise more efficient communication on some platforms, by allowing block transfers of larger data sets.
+
+[source,c++,Read write members,linenums]
+----
+template <class name,
+          class dataT,
+          size_t min_capacity = 0>
+class pipe {
+  template <pipe_property::writeable host_writeable>
+    static dataT* map(size_t requested_size, size_t &mapped_size);
```

Comment:
I wonder whether it would not be useful to have instead of `std::size_t` a numeric type as a template parameter, to avoid having 64 bit counters for a pipe of size 1 for example...

Reply:
Tracked for future in https://github.com/intel/llvm/issues/832

Reply:
Tracked for future in https://github.com/intel/llvm/issues/832